### PR TITLE
[ROCm] Expose kFp8DynamicTokenSym on AITER PTPC linears

### PR DESCRIPTION
**Pull request by rebklee**
## Purpose
`FP8ScaledMMLinearKernel.apply_weights` already no-ops in-kernel quant on a matching `QuantizedActivation`. The AITER PTPC kernels never exposed `input_quant_key()`, so `expose_input_quant_key` left the layer unmarked and producers could not skip `dynamic_per_token_scaled_quant`.

Return `kFp8DynamicTokenSym` from `input_quant_key()` on the AITER PTPC kernels (AiterPreshuffledPerTokenFp8ScaledMMLinearKernel, AiterPerTokenFp8ScaledMMLinearKernel, AiterHipbMMPerTokenFp8ScaledMMLinearKernel).

`Fp8PtpcOnlineLinearMethod` and `QuarkW8A8Fp8` call `expose_input_quant_key` after kernel init.

Not a duplicate. [#44260](https://github.com/vllm-project/vllm/pull/44260) landed the QuantizedActivation contract and CUDA static-FP8 consumers; AITER PTPC kernels still return None, so expose_input_quant_key never marks the layer. [#44132](https://github.com/vllm-project/vllm/pull/44132) added Fp8PtpcOnlineLinearMethod without that call. [#33773](https://github.com/vllm-project/vllm/pull/33773) landed the GEMMs, not the fusion key.

## Test Plan

Extended `tests/fusion/test_quant_activation_contract.py` with the three AITER PTPC kernels.

Added `test_quark_fp8_ptpc_exposes_kernel_input_quant_key`: `QuarkW8A8Fp8.create_weights` copies the kernel's `input_quant_key` onto the layer.

- [x] `pytest tests/fusion/test_quant_activation_contract.py --noconftest`
- [x] `pytest tests/quantization/test_quark.py::test_quark_fp8_ptpc_exposes_kernel_input_quant_key`

## Test Result

Image: `vllm/vllm-openai-rocm:nightly-1dc464d42681d22f38caf1fdc1eb632dc4421c45`

- `tests/fusion/test_quant_activation_contract.py` — 4 passed
- `tests/quantization/test_quark.py::test_quark_fp8_ptpc_exposes_kernel_input_quant_key` — 1 passed (`VLLM_ROCM_USE_AITER=1`)

### CODE REVIEW

File: tests/fusion/test_quant_activation_contract.py [hunk 1]
```diff
@@ -19,6 +19,11 @@
     FlashInferCutlassNvFp4LinearKernel,
     FlashInferTrtllmNvFp4LinearKernel,
 )
+from vllm.model_executor.kernels.linear.scaled_mm.aiter import (
```
Comment (okorzh-amd) [inline tests/fusion/test_quant_activation_contract.py:22]: PR description is missing test results

File: vllm/model_executor/layers/quantization/online/fp8.py [hunk 2]
```diff
@@ -22,6 +22,10 @@
     CutlassFP8ScaledMMLinearKernel,
     MarlinFP8ScaledMMLinearKernel,
 )
+from vllm.model_executor.layers.fusion.quant_activation import (
```
Comment (okorzh-amd) [inline vllm/model_executor/layers/quantization/online/fp8.py]: Blocker — lint. This block must come after the two layers.fused_moe imports (line 29-32); isort orders fused_moe < fusion. Repo's pinned ruff 0.14.0 reports I001 on this file; base commit passes. ruff check --fix does exactly this move. Note pre-commit never ran on this PR

File: vllm/model_executor/layers/quantization/online/fp8.py [hunk 3]
```diff
@@ -401,18 +405,21 @@ def process_weights_after_loading(self, layer: Module) -> None:
         replace_parameter(layer, "weight_scale", weight_scale)
 
         self.fp8_linear.process_weights_after_loading(layer)
+        expose_input_quant_key(layer, self.fp8_linear)
```
Comment (okorzh-amd) [inline vllm/model_executor/layers/quantization/online/fp8.py:408]: QuarkW8A8Fp8 reaches the same three kernels — quark_w8a8_fp8.py:60-74 builds kFp8DynamicTokenSym + kFp8StaticChannelSym, passes them to init_fp8_linear_kernel at :194, and the ROCm list at kernels/linear/__init__.py:412-420 puts the AITER PTPC kernels first — but
quark_w8a8_fp8.py:138 has no equivalent call. CompressedTensorsW8A8Fp8 already does (:146), so quark ends up the odd one out silently. Please add it.

File: vllm/model_executor/layers/quantization/online/fp8.py [hunk 4]
```diff
@@ -401,18 +405,21 @@ def process_weights_after_loading(self, layer: Module) -> None:
...
 
         layer._already_called_process_weights_after_loading = True
 
     def apply(
         self,
         layer: torch.nn.Module,
-        x: torch.Tensor,
+        x: torch.Tensor | QuantizedActivation,
```
Comment (okorzh-amd) [inline vllm/model_executor/layers/quantization/online/fp8.py]: Contravariant, so mypy is fine — but LinearMethodBase.apply still declares x: torch.Tensor (linear.py:158), and the nearest precedent ModelOptNvFp4LinearMethod keeps torch.Tensor (modelopt.py:1238) despite calling expose_input_quant_key at :1200. Worth widening the base in the same PR, or matching modelopt. (The runtime isinstance guard is needed either way.)

File: vllm/model_executor/layers/quantization/online/fp8.py [hunk 5]
```diff
@@ -401,18 +405,21 @@ def process_weights_after_loading(self, layer: Module) -> None:
...
         bias: torch.Tensor | None = None,
     ) -> torch.Tensor:
         # if batch invariant mode is enabled dequant
-        if envs.VLLM_BATCH_INVARIANT and not isinstance(
-            self.fp8_linear, CutlassFP8ScaledMMLinearKernel
+        if (
+            envs.VLLM_BATCH_INVARIANT
+            and not isinstance(self.fp8_linear, CutlassFP8ScaledMMLinearKernel)
+            and not isinstance(x, QuantizedActivation)
```
Comment (okorzh-amd) [inline vllm/model_executor/layers/quantization/online/fp8.py:422]: The guard is load-bearing — without it line 425's layer.weight.to(x.dtype) hits a dataclass with no .dtype. But the branch can't work with any AITER PTPC kernel regardless: line 407 lets the kernel rewrite the weight (aiter.py:366-374 transposes to [N,K]; :200-211 transposes and shuffles), so the [K,N] × [N,1] dequant only broadcasts when K == N. Selection isn't BI-aware for FP8 (only NVFP4 is, kernels/linear/__init__.py:1019-1050). Suggests VLLM_BATCH_INVARIANT + ROCm + AITER + online PTPC has never been run — forcing a
deterministic kernel at selection time would delete this branch and the question with it.

File: vllm/model_executor/kernels/linear/scaled_mm/aiter.py [hunk 6]
```diff
@@ -149,6 +151,14 @@ def is_supported(
             return False, "requires aiter library to be installed."
         return True, None
 
+    def input_quant_key(self) -> QuantKey | None:
+        # Advertise per-token FP8 so KDA gated-RMSNorm fusion can skip
```
Comment (okorzh-amd) [inline vllm/model_executor/kernels/linear/scaled_mm/aiter.py]: That fusion is in #54254, not in this tree. A reader grepping for it finds only the vendored KDA kernels under vllm/models/kimi_k3/, or the in-tree AiterRMSNormGatedFp8GroupQuantPattern (rocm_aiter_fusion.py:454-467), which is GroupShape(1,128) — a different key that as_quantized_activation would reject. Prefer environment-independent wording.

File: vllm/model_executor/kernels/linear/scaled_mm/aiter.py [hunk 7]
```diff
@@ -241,6 +251,11 @@ def is_supported(
 
         return True, None
 
+    def input_quant_key(self) -> QuantKey | None:
+        # Same PTPC consume ABI as the bpreshuffle kernel. HIPBMM stays off
```
Comment (okorzh-amd) [inline vllm/model_executor/kernels/linear/scaled_mm/aiter.py]: grep -rn 'in this harness' vllm/ returns only this line. Whether HIPBMM is on is a user env-var choice (VLLM_ROCM_USE_AITER_LINEAR_HIPBMM, see aiter.py:238-241), not a property of a bench. Also refer to the sibling by class name (AiterPreshuffledPerTokenFp8ScaledMMLinearKernel), not the aiter op name.

File: tests/fusion/test_quant_activation_contract.py [hunk 8]
```diff
@@ -46,6 +51,9 @@
     CutlassFP8ScaledMMLinearKernel,
     FlashInferFP8ScaledMMLinearKernel,
     FlashInferCutlassNvFp4LinearKernel,
+    AiterHipbMMPerTokenFp8ScaledMMLinearKernel,
+    AiterPreshuffledPerTokenFp8ScaledMMLinearKernel,
+    AiterPerTokenFp8ScaledMMLinearKernel,
```
Comment (okorzh-amd) [inline tests/fusion/test_quant_activation_contract.py:56]: All four tests in this file are set/attribute assertions over torch.zeros — zero GEMM execution.

File: tests/fusion/test_quant_activation_contract.py (hunk 1 above)
Comment (rebklee) [reply tests/fusion/test_quant_activation_contract.py:22]: Pasted in Test Result.

File: vllm/model_executor/layers/quantization/online/fp8.py (hunk 2 above)
Comment (rebklee) [reply vllm/model_executor/layers/quantization/online/fp8.py]: Fixed in 8598ebf.

File: vllm/model_executor/layers/quantization/online/fp8.py (hunk 3 above)
Comment (rebklee) [reply vllm/model_executor/layers/quantization/online/fp8.py:408]: Added on QuarkW8A8Fp8 after init_fp8_linear_kernel, same as CompressedTensorsW8A8Fp8. Covered by test_quark_fp8_ptpc_exposes_kernel_input_quant_key.

File: vllm/model_executor/layers/quantization/online/fp8.py (hunk 4 above)
Comment (rebklee) [reply vllm/model_executor/layers/quantization/online/fp8.py]: Matched ModelOpt, and apply stays x: torch.Tensor. Left LinearMethodBase alone. Kept the isinstance guard.

File: tests/fusion/test_quant_activation_contract.py (hunk 8 above)
Comment (rebklee) [reply tests/fusion/test_quant_activation_contract.py:56]: This PR only advertises the key; unfused PTPC still quantizes inside the linear. A GEMM of QuantizedActivation vs in-kernel quant belongs with a producer or a follow-up. The Quark test is wiring (create_weights), not a GEMM.

File: vllm/model_executor/kernels/linear/scaled_mm/aiter.py (hunk 6 above)
Comment (rebklee) [reply vllm/model_executor/kernels/linear/scaled_mm/aiter.py]: Comments have been reworded.

File: vllm/model_executor/kernels/linear/scaled_mm/aiter.py (hunk 7 above)
Comment (rebklee) [reply vllm/model_executor/kernels/linear/scaled_mm/aiter.py]: HIPBMM comment now names AiterPreshuffledPerTokenFp8ScaledMMLinearKernel.

File: vllm/model_executor/layers/quantization/online/fp8.py (hunk 5 above)
Comment (rebklee) [reply vllm/model_executor/layers/quantization/online/fp8.py:422]: Kept the isinstance guard so a QuantizedActivation does not hit x.dtype. Not changing FP8 kernel selection in this PR, and VLLM_BATCH_INVARIANT + ROCm + AITER + online PTPC is out of scope.

**Review by dllehr-amd — APPROVED** [review]
Thanks @rebklee

**Comment by rebklee** [comment]
@AndreasKaratzas this is approved and passing CI on `2f78ba7`. Could you please take a look and add `ready` if it looks good? Thanks!
